samcli/lib/utils/git_repo.py

```diff
@@ -0,0 +1,188 @@
+        """
+        creates a local clone of this Git repository
+
+        Parameters
+        ----------
+        clone_dir: Path
+            The directory to create the local clone inside
+        clone_name: Optional[str]
+            The dirname of the local clone, if not provided, repository's name (self._name) will be used
+        replace_existing: bool
+            Whether to replace the current local clone directory if already exists or not
+
+        Returns
+        -------
+            The path of the created local clone
+
+        Raises
+        ------
+        OSError:
+            when file management errors like unable to mkdir, copytree, rmtree ...etc
+        CloneRepoException:
+            General errors like for example; if an error occurred while running `git clone`
+            or if the local_clone already exists and replace_existing is not set
+        CloneRepoUnstableStateException:
+            when reaching unstable state, for example with replace_existing flag set, unstable state can happen
+            if removed the current local clone but failed to copy the new one from the temp location to the destination
+        """
+        if not self._auto_clone:  # Unit test escape hatch
+            return None  # type: ignore
+        GitRepo._ensure_clone_directory_exists(clone_dir=clone_dir)
+        # clone to temp then move to the destination(repo_local_path)
+        with osutils.mkdir_temp(ignore_errors=True) as tempdir:
+            try:
+                clone_name = clone_name if clone_name else self._name
+                temp_path = os.path.normpath(os.path.join(tempdir, clone_name))
+                git_executable: str = GitRepo._git_executable()
+                LOG.info("\nCloning from %s", self._url)
+                subprocess.check_output(
+                    [git_executable, "clone", self._url, clone_name],
+                    cwd=tempdir,
+                    stderr=subprocess.STDOUT,
+                )
+                self._local_path = self._persist_local_repo(temp_path, clone_dir, clone_name, replace_existing)
+                return self._local_path
+            except OSError as ex:
+                LOG.warning("WARN: Could not clone repo %s", self._url, exc_info=ex)
+                raise
+            except subprocess.CalledProcessError as clone_error:
+                output = clone_error.output.decode("utf-8")
+                if "not found" in output.lower():
+                    LOG.warning("WARN: Could not clone repo %s", self._url, exc_info=clone_error)
+                raise CloneRepoException from clone_error
+            finally:
+                self._clone_attempted = True
+
+    @staticmethod
+    def _persist_local_repo(temp_path: str, dest_dir: Path, dest_name: str, replace_existing: bool) -> Path:
+        dest_path = os.path.normpath(dest_dir.joinpath(dest_name))
+        try:
+            if Path(dest_path).exists():
```

Comment:
Could we optimize this by checking this before cloning and directly cloning into the right directory instead of copying every time?

Reply:
Ok, we may want to keep the old clone if the new clone fails I guess